### PR TITLE
Impl BitAnd, BitAndAssign for bitmask

### PR DESCRIPTION
**Pull request by Riey**
This is useful when check specific mask.

```rust
let key_mask: KeyButMask = some_how();

if (key_mask & KeyButMask::SHIFT) != KeyButMask::SHIFT {
    // Do some stuff
} 
```

### CODE REVIEW

File: x11rb-protocol/src/x11_utils.rs [hunk 1]
```diff
@@ -607,6 +607,34 @@ macro_rules! bitmask_binop {
...
+            }
+        }
+        impl core::ops::BitAndAssign<$t> for $u {
+            fn bitand_assign(&mut self, other: $t) {
+                *self &= Self::from(other)
+            }
+        }
+        impl core::ops::BitAndAssign<$u> for $t {
+            fn bitand_assign(&mut self, other: $u) {
+                *self = *self & Self::from(other)
```
Comment (notgull) [inline x11rb-protocol/src/x11_utils.rs:635]: Nitpick: Wouldn't it be better to use `&=` here, semantically?

**Review by notgull — COMMENTED** [review]
LGTM aside from the nitpick

File: x11rb-protocol/src/x11_utils.rs (hunk 1 above)
Comment (Riey) [reply x11rb-protocol/src/x11_utils.rs:635]: I just C-V above `BitOr` codes should I edit both?

Comment (Riey) [reply x11rb-protocol/src/x11_utils.rs:635]: `*self &= ` Can't be used because that operator'll call `BitAndAssign` which cause infinite loop.

Comment (notgull) [reply x11rb-protocol/src/x11_utils.rs:635]: I meant over `self.0` or whatever this type is a newtype around.

**Review by notgull — APPROVED** [review]

**Comment by psychon** [comment]
```rust
(key_mask & KeyButMask::SHIFT) != KeyButMask::SHIF
```
Should we provide some explicit helper to make the above more readable?

Looking at https://docs.rs/bitmask/latest/bitmask/macro.bitmask.html, I guess at least `contains` can be useful. Providing "everything" could be a bit excessive... (I am looking at you, compile times)

Oh and: Do we want this instead of what this PR added or in addition to it?
```diff
diff --git a/x11rb-protocol/src/x11_utils.rs b/x11rb-protocol/src/x11_utils.rs
index 55aba54..c0982cc 100644
--- a/x11rb-protocol/src/x11_utils.rs
+++ b/x11rb-protocol/src/x11_utils.rs
@@ -575,7 +575,7 @@ impl<T: Serialize> Serialize for [T] {
     }
 }
 
-// This macro is used by the generated code to implement `std::ops::BitOr` and
+// This macro is used by the generated code to implement e.g. `std::ops::BitOr` and
 // `std::ops::BitOrAssign`.
 macro_rules! bitmask_binop {
     ($t:ty, $u:ty) => {
@@ -607,6 +607,15 @@ macro_rules! bitmask_binop {
                 *self = *self | Self::from(other)
             }
         }
+        impl $t {
+            //! Check if this object has all bits set that are also set in `flag`.
+            //!
+            //! `flag` can be a single enum variant or a whole other mask.
+            pub fn contains(self, flag: impl Into<$u>) -> bool {
+                let flag = flag.into();
+                (<$u>::from(self) & flag) == flag
+            }
+        }
     };
 }
 
```
(I will not turn the above into a PR unless someone wants me to)

**Comment by Riey** [comment]
@psychon I personally want many of `bitflags` methods. But I don't know how many types will be affected.

**Comment by psychon** [comment]
Okay... I opened #769 which contains basically the above. While doing this, I noticed that e.g. `EventMask | EventMask` was missing, so I added `impl`s for that as well.

**Pull request merged**